blackjax/mcmc/ghmc.py

```diff
@@ -0,0 +1,209 @@
+class GHMCState(NamedTuple):
+    position: PyTree
+    momentum: PyTree
+    potential_energy: float
+    potential_energy_grad: PyTree
+    slice: float
+
+
+class GHMCInfo(NamedTuple):
+    momentum: PyTree
+    acceptance_probability: float
+    is_accepted: bool
+    is_divergent: bool
+    energy: float
+    proposal: proposal.Proposal
+
+
+def init(
+    rng_key: PRNGKey,
+    position: PyTree,
+    logprob_fn: Callable,
+    logprob_grad_fn: Optional[Callable] = None,
+):
+    def potential_fn(x):
+        return -logprob_fn(x)
+
+    if logprob_grad_fn:
+        potential_energy_grad = jax.tree_map(
+            lambda g: -1.0 * g, logprob_grad_fn(position)
+        )
+        potential_energy = potential_fn(position)
+
+    else:
+        potential_energy, potential_energy_grad = jax.value_and_grad(potential_fn)(
+            position
+        )
+
+    p, unravel_fn = ravel_pytree(position)
+    key_mometum, key_slice = jax.random.split(rng_key)
+    momentum = unravel_fn(jax.random.normal(key_mometum, p.shape))
+    slice = jax.random.uniform(key_slice, minval=-1.0, maxval=1.0)
+
+    return GHMCState(position, momentum, potential_energy, potential_energy_grad, slice)
+
+
+def kernel(
+    noise_gn: Callable = lambda _: 0.0,
+    divergence_threshold: float = 1000,
+):
+    def kinetic_energy_fn(momentum):
+        momentum, _ = ravel_pytree(momentum)
+        return 0.5 * jnp.dot(momentum, momentum)
+
+    def one_step(
+        rng_key: PRNGKey,
+        state: GHMCState,
+        logprob_fn: Callable,
+        step_size: PyTree,  # float,
+        alpha: float,
+        delta: float,
+        logprob_grad_fn: Optional[Callable] = None,
+    ):
+        def potential_fn(x):
+            return -logprob_fn(x)
+
+        proposal_generator = persistent_ghmc_proposal(
+            potential_fn,
+            kinetic_energy_fn,
+            step_size,
+            divergence_threshold,
+            logprob_grad_fn,
+        )
+
+        key_momentum, key_noise = jax.random.split(rng_key)
+        position, momentum, potential_energy, potential_energy_grad, slice = state
+        momentum, unravel_fn = ravel_pytree(momentum)
+        # New momentum is persistent
+        momentum = unravel_fn(
+            momentum * jnp.sqrt(1.0 - alpha)
+            + jnp.sqrt(alpha) * jax.random.normal(key_momentum, jnp.shape(momentum))
+        )
+        # Slice is non-reversible
+        slice = ((slice + 1.0 + delta + noise_gn(key_noise)) % 2) - 1.0
+
+        integrator_state = integrators.IntegratorState(
+            position, momentum, potential_energy, potential_energy_grad
+        )
+        proposal, info = proposal_generator(slice, integrator_state)
+        state = GHMCState(
+            proposal.position,
+            proposal.momentum,
+            proposal.potential_energy,
+            proposal.potential_energy_grad,
+            slice,
+        )
+
+        return state, info
+
+    return one_step
+
+
+def persistent_ghmc_proposal(
```

Comment:
Isn't that essentially the same as HMC's? We should re-use the code as much as possible: it simplified the codebase and make obvious the connections between the algorithms when there are some. In this case GHMC is "just" HMC with a different momentum sampling and acceptance schemes. We should make this obvious looking at the code.